transports/quic/src/transport.rs

```diff
@@ -0,0 +1,350 @@
+            libp2p_core::PublicKey::from_protobuf_encoding(peer_id.digest()).map_err(|_| ())?;
+        let public_key = if let libp2p_core::PublicKey::Ed25519(public_key) = public_key {
+            public_key.encode()
+        } else {
+            return Err(());
+        };
+        let public_key = PublicKey::from_bytes(&public_key).map_err(|_| ())?;
+        iter.next();
+        Some(public_key)
+    } else {
+        None
+    };
+
+    if iter.next().is_some() {
+        return Err(());
+    }
+
+    match (proto1, proto2, proto3) {
+        (Protocol::Ip4(ip), Protocol::Udp(port), Protocol::Quic) => {
+            Ok((SocketAddr::new(ip.into(), port), peer_id))
+        }
+        (Protocol::Ip6(ip), Protocol::Udp(port), Protocol::Quic) => {
+            Ok((SocketAddr::new(ip.into(), port), peer_id))
+        }
+        _ => Err(()),
+    }
+}
+
+/// Turns an IP address and port into the corresponding QUIC multiaddr.
+pub(crate) fn socketaddr_to_multiaddr(socket_addr: &SocketAddr) -> Multiaddr {
+    Multiaddr::empty()
+        .with(socket_addr.ip().into())
+        .with(Protocol::Udp(socket_addr.port()))
+        .with(Protocol::Quic)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn multiaddr_to_udp_conversion() {
+        use std::net::{IpAddr, Ipv4Addr, Ipv6Addr};
+
+        assert!(
+            multiaddr_to_socketaddr(&"/ip4/127.0.0.1/udp/1234".parse::<Multiaddr>().unwrap())
+                .is_err()
+        );
+
+        assert_eq!(
+            multiaddr_to_socketaddr(
+                &"/ip4/127.0.0.1/udp/12345/quic"
+                    .parse::<Multiaddr>()
+                    .unwrap()
+            ),
+            Ok((
+                SocketAddr::new(IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1)), 12345,),
+                None
+            ))
+        );
+        assert_eq!(
+            multiaddr_to_socketaddr(
+                &"/ip4/255.255.255.255/udp/8080/quic"
+                    .parse::<Multiaddr>()
+                    .unwrap()
+            ),
+            Ok((
+                SocketAddr::new(IpAddr::V4(Ipv4Addr::new(255, 255, 255, 255)), 8080,),
+                None
+            ))
+        );
+        assert_eq!(
+            multiaddr_to_socketaddr(&"/ip6/::1/udp/12345/quic".parse::<Multiaddr>().unwrap()),
+            Ok((
+                SocketAddr::new(IpAddr::V6(Ipv6Addr::new(0, 0, 0, 0, 0, 0, 0, 1)), 12345,),
+                None
+            ))
+        );
+        assert_eq!(
+            multiaddr_to_socketaddr(
+                &"/ip6/ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff/udp/8080/quic"
+                    .parse::<Multiaddr>()
+                    .unwrap()
+            ),
+            Ok((
+                SocketAddr::new(
+                    IpAddr::V6(Ipv6Addr::new(
+                        65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535,
+                    )),
+                    8080,
+                ),
+                None
+            ))
+        );
```

Comment:
No tests for `/ipV/ADDR/udp/PORt/quic/p2p/PEERID`.

Reply:
Done